### ISSUE TITLE
Dealing with under/over - fitting based on epochs

### ISSUE THREAD

**Issue by Samek24** [DeepGraphLearning/torchdrug#81]
Hi,

On the property prediction tutorial, the model is trained on 100 epochs : 
![image](https://user-images.githubusercontent.com/49188271/160423448-f21e1424-1d07-4e97-b414-e3f25c9b1bb1.png)

Here 100 is arbitrary and it seems that the model is the one obtained after 100 epochs but there is no guarantee that the model has not overfitted (or underfitted) : did I understand right ?
Is it possible to have access to the loss at each epoch (for both training and validation) in order to control learning ?

Thank you !

**Comment by Oxer11** [comment]
Hi! Here we set num_epoch=100 for illustration and the final model is the one after 100 epochs.

In practice, you may want to have a look at the validation metric and save the model during training. We haven't implement this in core.Engine, but you can do this by using a loop outside the solver:
```
for i in range(0, num_epoch, step):
    solver.train(num_epoch = min(step, num_epoch - i))
    metric = solver.evaluate("valid")
    solver.save("model_epoch_%d.pth" % solver.epoch)
```
where the variable `step` controls the frequency you want to evaluate and save your model.

As for the training loss function, we will consider to return some related values in the core.Engine.train() method in the future version.

**Comment by Samek24** [comment]
If I understand correctly, in core.Engine, the arguments "valid_set" and "test_set" are not used during training (solver.train), but they are mandatory, right ?

Indeed we can use a for loop as you suggest, but in the code above I don't understand why the "min" is taken to choose "num_epoch". If we take for example num_epoch = 100 and step = 10, models obtained with num_epoch=1 to 10 will be evaluated and saved. 

Shouldn't we juste write:

`solver.train(num_epoch=i)` ? (and handling the case when i=0 ...)

**Comment by Oxer11** [comment]
Yes. "valid_set" and "test_set" are only used during evaluation. Although they're mandatory, you can still pass a `None` variable to it if you don't want to use the `evaluate()` method.

For your second question, it should be noted that the `num_epoch` argument in the `solver.train()` method represents the number of epochs you want to train in this call (not in total). You can check the implementation [here](https://torchdrug.ai/docs/_modules/torchdrug/core/engine.html#Engine.train). In the code snippet above, we want to train for `step` epochs for each iteration except the last one. In the last iteration, we only need to train `num_epoch - i` epochs, which is not more than `step`. So we need to take the `min(step, num_epoch-i)` as the argument.

If we take `num_epoch=45` and `step=10`, we will train for `10, 10, 10, 10, 5` epochs in the five iterations of the loop, repsectively.
